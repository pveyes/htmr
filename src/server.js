// @flow
import parse from 'posthtml-parser';
import React from 'react';
import { AllHtmlEntities as HtmlEntity } from 'html-entities';
import mapAttribute from './mapAttribute';

import type { NodeMap, ConvertedComponent } from './types';

// eslint-disable-next-line no-use-before-define
type Node = ElementNode | TextNode;

type TextNode = string;

type ElementNode = {
  tag: string,
  attrs: {
    [key: string]: string,
  },
  content: Array<Node>,
};

type Element = React$Element<*> | string;

function transform(node: Node, key: string, nodeMap: NodeMap): ?Element {
  if (typeof node === 'string') {
    // newline and space will be parsed as 'node' in posthtml-parser,
    // we can ignore it along with comment node
    const text = node.trim();
    if (/^<!--[\s\S]+-->/.test(text)) {
      return null;
    }
    if (text === '') {
      return node;
    }

    return HtmlEntity.decode(node);
  }

  const { tag, attrs, content } = node;
  const Component = nodeMap[tag] || tag;

  const props = Object.assign(
    {},
    mapAttribute(attrs),
    // always set key because it's possible the html source contains
    // multiple elements
    { key }
  );

  // style tag needs to preserve its children
  if (tag === 'style' && Component === tag) {
    props.dangerouslySetInnerHTML = { __html: content[0] };
    return React.createElement(tag, props, null);
  }

  // self closing component doesn't have children
  let children =
    content === undefined
      ? null
      : content
          .map((child, index) => {
            const childKey = `${key}.${index}`;
            return transform(child, childKey, nodeMap);
          })
          .filter(child => child !== null);

  if (children && children.length === 0) {
    children = null;
  }

  return React.createElement(Component, props, children);
}

function convertServer(
  html: string,
  nodeMap: NodeMap = {}
): ConvertedComponent {
  const ast = parse(html.trim());
  const components = ast
    .map((node, index) => transform(node, index.toString(), nodeMap))
    .filter(node => node !== null);

  if (components.length > 1) {
    return components;
  }

  return components[0];
}

module.exports = convertServer;
