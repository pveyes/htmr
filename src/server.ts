import React, { ReactNode } from 'react';
import { HtmrOptions, HTMLTags } from "./types";
import Parser from 'htmlparser2';
import { AllHtmlEntities as HtmlEntity } from 'html-entities';
import mapAttribute from './mapAttribute';

type HTMLNode = {
  type: 'tag' | 'style',
  name: HTMLTags,
  attribs: {
    [key: string]: any
  }
  children: Array<Node>
}

type TextNode = {
  type: 'text',
  data: string,
  parent?: HTMLNode
}

type Node = HTMLNode | TextNode;

const TABLE_ELEMENTS = ['table', 'tbody', 'thead', 'tfoot', 'tr'];

function transform(node: Node, key: string, options: HtmrOptions): ReactNode {
  const defaultTransform = options.transform._;

  switch (node.type) {
    case 'style':
    case 'tag': {
      const { name, attribs } = node;

      // decode all attribute value
      Object.keys(attribs).forEach(key => {
        attribs[key] = HtmlEntity.decode(attribs[key]);
      });

      const props = Object.assign(
        {},
        mapAttribute(attribs, options.preserveAttributes),
        { key }
      );

      // if the tags children should be set dangerously
      if (options.dangerouslySetChildren.indexOf(name) > -1) {
        const childNode = <TextNode>node.children[0];
        props.dangerouslySetInnerHTML = {
          __html: childNode.data.trim()
        };
        return React.createElement(name, props, null);
      }

      const childNodes = node.children
        .map((node, index) => transform(node, index.toString(), options))
        .filter(Boolean);

      // self closing component doesn't have children
      const children = childNodes.length === 0
        ? null
        : childNodes;

      const customElement = options.transform[name];
      if (customElement) {
        return React.createElement(customElement, props, children);
      }

      if (defaultTransform) {
        return defaultTransform(name, props, children);
      }

      return React.createElement(name, props, children);
    }
    case 'text': {
      let str = node.data;

      if (node.parent && TABLE_ELEMENTS.indexOf(node.parent.name) > -1) {
        str = str.trim();
        if (str === '') {
          return null;
        }
      }

      str = HtmlEntity.decode(str);

      return defaultTransform ? defaultTransform(str) : str;
    }
  }
}

export default function convertServer(html: string, options: Partial<HtmrOptions> = {}) {
  if (typeof html !== 'string') {
    throw new TypeError('Expected HTML string');
  }

  const opts: HtmrOptions = {
    transform: options.transform || {},
    preserveAttributes: options.preserveAttributes || [],
    dangerouslySetChildren: options.dangerouslySetChildren || ["style"],
  };

  const ast = Parser.parseDOM(html.trim(), {}) as Array<HTMLNode>;
  const components = ast.map((node, index) => transform(node, index.toString(), opts));

  if (components.length > 1) {
    return components;
  }

  return components[0];
}
