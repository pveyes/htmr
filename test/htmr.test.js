/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import { renderToStaticMarkup, renderToString } from 'react-dom/server';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import snapshot from 'jest-snapshot';
import diff from 'jest-diff';
import { html, oneLineTrim } from 'common-tags';

import htmrServer from '../src';
import htmrBrowser from '../src/index.browser';

describe('core', () => {
  test('it works', () => {
    testRender('<p>This is cool</p>');
  });

  test('throws if first argument is not a string', () => {
    const error = new Error('Expected HTML string');
    const fixtures = [null, [], {}, 1, true];

    // server & browser
    expect.assertions(fixtures.length * 2);

    fixtures.forEach((fixture) => {
      expect(() => htmrServer(fixture)).toThrow(error);
      expect(() => htmrBrowser(fixture)).toThrow(error);
    });
  });

  test('can render self closing component', () => {
    testRender('<img src="https://www.google.com/logo.png" />');
    testRender('<hr>');
    testRender('<br />');
  });

  test('can render sibling element as multiple children', () => {
    testRender('<p>Multi</p><p>Component</p>');
  });

  test('can render element inside text node', () => {
    testRender('what are <strong>you</strong> doing?');
  });

  test('ignore comment', () => {
    testRender('<!-- comment should be ignored--><div>no comment</div>');
  });

  test('ignore multiline comment', () => {
    testRender(oneLineTrim`
      <!--<div>\n<p>multiline</p> \t</div>-->
      <li>no multiline comment</li>
    `);
  });
});

describe('attributes', () => {
  test('correctly map HTML attributes to react props', () => {
    testRender('<label class="input-text" for="name"></label>');
    testRender(
      '<div id="test" data-type="calendar" aria-describedby="info" spellcheck="true" contenteditable></div>'
    );
    testRender('<link xml:lang="en" xlink:actuate="other" />');
    testRender(oneLineTrim`
      <svg viewbox="0 0 24 24" fill-rule="evenodd" color-interpolation-filters="sRGB">
        <path fill="#ffa0"></path>
      </svg>
    `);
    testRender('<img srcset="https://img.src" crossorigin="true" />');
    testRender('<iframe srcdoc="<p>html</p>" allowfullscreen></iframe>');
    testRender(
      '<input autocomplete="on" autofocus readonly="readonly" maxlength="10" />'
    );
    testRender('<button accesskey="s">Stress reliever</button>');
    testRender('<time datetime="2018-07-07">July 7</time>');
  });

  // https://github.com/pveyes/htmr/issues/103
  test('correctly handle boolean attributes', () => {
    const { container } = render(htmrBrowser('<iframe allowfullscreen />'));
    expect(
      container.querySelector('iframe').getAttribute('allowfullscreen')
    ).toEqual('');
  });

  test('correctly convert multiple style values', () => {
    testRender('<ul style="margin: 0 auto; padding: 0 10px"></ul>');
    testRender('<span style="font-size: 12"></span>');
    testRender(
      '<div style="background-image:url(https://d1nabgopwop1kh.cloudfront.net/xx);"></div>'
    );
  });

  test('can handle css vendor prefixes', () => {
    testRender(
      '<i style="-ms-text-size-adjust: 100%; -webkit-text-size-adjust: 100%">prefix</i>'
    );
  });

  test('decode HTML entities inside style declaration', () => {
    testRender(
      '<em style="font-family: Consolas, &quot;Liberation Mono &quot;"></em>'
    );
  });

  test('ignore invalid style', () => {
    testRender('<div style="TITLE_2">Explore Categories</div>');
    testRender('<div style="TITLE_2; color:\'red\'">Explore Categories</div>');
  });
});

describe('encoding', () => {
  test('unescape html entities', () => {
    testRender('<blockquote class="entities">&amp; and &</blockquote>');
  });

  test('decode html entities on defaultMap', () => {
    testRender('<div class="entities">&amp; and &</div>', {
      transform: {
        _: (node, props, children) => {
          if (typeof props === 'undefined') {
            return node;
          }

          return <p>{children}</p>;
        },
      },
    });
  });

  test('decode html attributes', () => {
    testRender('<a href="https://www.google.com/?a=b&amp;c=d">test</a>');
  });

  test('dangerously rendered script tag is not encoded', () => {
    const html = `
      <script data-cfasync="false" type="application/json">
        {
          "key": "value"
        }
      </script>
    `.trim();

    testRender(html, { dangerouslySetChildren: ['script'] });
  });
});

describe('options', () => {
  describe('transform', () => {
    test('custom component', () => {
      const Section = (props) => {
        const { children, 'data-heading': heading, ...rest } = props;
        return (
          <section {...rest} className="css-x243s">
            <h2>{heading}</h2>
            {children}
          </section>
        );
      };

      testRender(
        '<p data-custom="true" data-heading="Recommended">Custom component</p>',
        { transform: { p: Section } }
      );
    });

    test('default mapping', () => {
      let i = 0;
      const defaultMap = (node, props, children) => {
        if (typeof props === 'undefined') {
          // we need to add key for elements inside array
          return <span key={i++}>{node}</span>;
        }

        return <div {...props}>{children}</div>;
      };

      testRender('<article><p>Default mapping</p></article>', {
        transform: { _: defaultMap },
      });
    });
  });

  describe('preserveAttributes', () => {
    test('allow preserve some attributes', () => {
      testRender(
        html`
          <div ng-if="x">
            <div tv-abc="d" tv-xxx="y"></div>
          </div>
        `,
        { preserveAttributes: ['ng-if', new RegExp('tv-')] }
      );
    });
  });

  describe('dangerouslySetChildren', () => {
    test('preserve the content of style tag by default', () => {
      testRender(oneLineTrim`
        <style>
          ul > li {
            list-style: none
          }
    
          div[data-id="test"]:not(.y) {
            display: none;
          }
        </style>
      `);
    });

    test('should dangerously set html for required tags', () => {
      const html = oneLineTrim`
        <pre>
          &lt;a href=&quot;/&quot;&gt;Test&lt;/a&gt;
        </pre>
      `;

      testRender(html, { dangerouslySetChildren: ['pre'] });
    });

    test('no dangerously render script tag', () => {
      testRender(html`
        <script data-cfasync="false" type="text/javascript">
          var gtm4wp_datalayer_name = 'dataLayer';
          var dataLayer = dataLayer || [];
          dataLayer.push({
            pagePostType: 'post',
            pagePostType2: 'single-post',
            pageCategory: ['kalender-cuti'],
            pagePostAuthor: 'Candra Alif Irawan',
          });
        </script>
      `);
    });

    test('dangerously render empty script tag', () => {
      testRender('<script type="text/javascript"></script>', {
        dangerouslySetChildren: ['script'],
      });
    });
  });
});

describe('whitespace', () => {
  test('allow whitespace only text nodes between elements', () => {
    testRender('<span>Hello</span> <span>World</span>');
  });

  test('allow newline only text node between elements', () => {
    testRender('<pre><span>Hello</span>\n<span>World</span></pre>');
  });

  test('remove whitespace on table elements', () => {
    testRender(html`
      <table>
        <tbody>
          <tr>
            <th>title</th>
          </tr>
          <tr>
            <td>entry</td>
          </tr>
        </tbody>
      </table>
    `);
  });
});

expect.extend({
  toRenderConsistently({ server, browser }, html) {
    const serverRender = renderer.create(server);
    const browserRender = renderer.create(browser);

    const serverHtml = snapshot.utils.serialize(serverRender);
    const browserHtml = snapshot.utils.serialize(browserRender);

    const diffString = diff(serverHtml, browserHtml, {
      expand: this.expand,
      aAnnotation: 'Server render',
      bAnnotation: 'Browser render',
    });
    const pass = serverHtml === browserHtml;
    let messageExpectation;

    if (pass) {
      messageExpectation =
        'Expected server rendered HTML to not equal browser rendered HTML';
    } else {
      messageExpectation =
        'Expected server rendered HTML to equal browser rendered HTML';
    }

    const message = () =>
      messageExpectation +
      '\n\n' +
      'Server render:\n' +
      `  ${this.utils.printExpected(serverHtml)} \n` +
      'Browser render:\n' +
      `  ${this.utils.printReceived(browserHtml)} \n` +
      (diffString ? `\n\nDifference: \n\n${diffString} ` : '');

    return { message, pass };
  },
});

/**
 * Test utilities
 */

function testRender(html, options) {
  const server = htmrServer(html, options);
  const browser = htmrBrowser(html, options);

  expect({ server, browser }).toRenderConsistently(html);

  // assert SSR
  expect(() => renderToString(server)).not.toThrow();
  expect(() => renderToStaticMarkup(server)).not.toThrow();

  // assert CSR
  const el = document.createElement('div');
  try {
    document.body.appendChild(el);
    expect(() => {
      ReactDOM.render(browser, el);
    }).not.toThrow();
  } finally {
    document.body.removeChild(el);
  }

  // assert snapshot, doesn't matter from server or browser
  // because we've already done assert equal between them
  expect(renderer.create(server)).toMatchSnapshot();
}
