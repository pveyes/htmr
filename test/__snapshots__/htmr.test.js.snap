// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attributes can handle css vendor prefixes 1`] = `
<i
  style={
    Object {
      "WebkitTextSizeAdjust": "100%",
      "msTextSizeAdjust": "100%",
    }
  }
>
  prefix
</i>
`;

exports[`attributes correctly convert multiple style values 1`] = `
<ul
  style={
    Object {
      "margin": "0 auto",
      "padding": "0 10px",
    }
  }
/>
`;

exports[`attributes correctly convert multiple style values 2`] = `
<span
  style={
    Object {
      "fontSize": 12,
    }
  }
/>
`;

exports[`attributes correctly convert multiple style values 3`] = `
<div
  style={
    Object {
      "backgroundImage": "url(https://d1nabgopwop1kh.cloudfront.net/xx)",
    }
  }
/>
`;

exports[`attributes correctly map HTML attributes to react props 1`] = `
<label
  className="input-text"
  htmlFor="name"
/>
`;

exports[`attributes correctly map HTML attributes to react props 2`] = `
<div
  aria-describedby="info"
  contentEditable="true"
  data-type="calendar"
  id="test"
  spellCheck="true"
/>
`;

exports[`attributes correctly map HTML attributes to react props 3`] = `
<link
  xlinkActuate="other"
  xmlLang="en"
/>
`;

exports[`attributes correctly map HTML attributes to react props 4`] = `
<svg
  colorInterpolationFilters="sRGB"
  fillRule="evenodd"
  viewBox="0 0 24 24"
>
  <path
    fill="#ffa0"
  />
</svg>
`;

exports[`attributes correctly map HTML attributes to react props 5`] = `
<img
  crossOrigin="true"
  srcSet="https://img.src"
/>
`;

exports[`attributes correctly map HTML attributes to react props 6`] = `
<iframe
  allowFullScreen={true}
  srcDoc="<p>html</p>"
/>
`;

exports[`attributes correctly map HTML attributes to react props 7`] = `
<input
  autoComplete="on"
  autoFocus={true}
  maxLength="10"
  readOnly={true}
/>
`;

exports[`attributes correctly map HTML attributes to react props 8`] = `
<button
  accessKey="s"
>
  Stress reliever
</button>
`;

exports[`attributes correctly map HTML attributes to react props 9`] = `
<time
  dateTime="2018-07-07"
>
  July 7
</time>
`;

exports[`attributes correctly map HTML attributes to react props 10`] = `
<img
  alt="alt"
  className=""
/>
`;

exports[`attributes decode HTML entities inside style declaration 1`] = `
<em
  style={
    Object {
      "fontFamily": "Consolas, \\"Liberation Mono \\"",
    }
  }
/>
`;

exports[`attributes handle CSS custom properties correctly 1`] = `
<div
  style={
    Object {
      "--color": "red",
    }
  }
>
  Explore Categories
</div>
`;

exports[`attributes handle CSS custom properties correctly 2`] = `
<div
  style={
    Object {
      "--color": "var(--syntax-text)",
    }
  }
>
  Explore Categories
</div>
`;

exports[`attributes ignore invalid style 1`] = `
<div
  style={Object {}}
>
  Explore Categories
</div>
`;

exports[`attributes ignore invalid style 2`] = `
<div
  style={
    Object {
      "color": "\\"red\\"",
    }
  }
>
  Explore Categories
</div>
`;

exports[`core can render element inside text node 1`] = `
Array [
  "what are ",
  <strong>
    you
  </strong>,
  " doing?",
]
`;

exports[`core can render self closing component 1`] = `
<img
  src="https://www.google.com/logo.png"
/>
`;

exports[`core can render self closing component 2`] = `<hr />`;

exports[`core can render self closing component 3`] = `<br />`;

exports[`core can render sibling element as multiple children 1`] = `
Array [
  <p>
    Multi
  </p>,
  <p>
    Component
  </p>,
]
`;

exports[`core ignore comment 1`] = `
<div>
  no comment
</div>
`;

exports[`core ignore multiline comment 1`] = `
<li>
  no multiline comment
</li>
`;

exports[`core it works 1`] = `
<p>
  This is cool
</p>
`;

exports[`encoding dangerously rendered script tag is not encoded 1`] = `
<script
  dangerouslySetInnerHTML={
    Object {
      "__html": "{
          \\"key\\": \\"value\\"
        }",
    }
  }
  data-cfasync="false"
  type="application/json"
/>
`;

exports[`encoding decode html attributes 1`] = `
<a
  href="https://www.google.com/?a=b&c=d"
>
  test
</a>
`;

exports[`encoding decode html entities on defaultMap 1`] = `
<p>
  & and &
</p>
`;

exports[`encoding unescape html entities 1`] = `
<blockquote
  className="entities"
>
  & and &
</blockquote>
`;

exports[`options dangerouslySetChildren dangerously render empty script tag 1`] = `
<script
  type="text/javascript"
/>
`;

exports[`options dangerouslySetChildren no dangerously render script tag 1`] = `
<script
  data-cfasync="false"
  type="text/javascript"
>
  
  var gtm4wp_datalayer_name = 'dataLayer';
  var dataLayer = dataLayer || [];
  dataLayer.push({
    pagePostType: 'post',
    pagePostType2: 'single-post',
    pageCategory: ['kalender-cuti'],
    pagePostAuthor: 'Candra Alif Irawan',
  });

</script>
`;

exports[`options dangerouslySetChildren preserve the content of style tag by default 1`] = `
<style
  dangerouslySetInnerHTML={
    Object {
      "__html": "ul > li {list-style: none}div[data-id=\\"test\\"]:not(.y) {display: none;}",
    }
  }
/>
`;

exports[`options dangerouslySetChildren should dangerously set html for required tags 1`] = `
<pre
  dangerouslySetInnerHTML={
    Object {
      "__html": "&lt;a href=&quot;/&quot;&gt;Test&lt;/a&gt;",
    }
  }
/>
`;

exports[`options preserveAttributes allow preserve some attributes 1`] = `
<div
  ng-if="x"
>
  
  
  <div
    tv-abc="d"
    tv-xxx="y"
  />
  

</div>
`;

exports[`options transform custom component 1`] = `
<section
  className="css-x243s"
  data-custom="true"
>
  <h2>
    Recommended
  </h2>
  Custom component
</section>
`;

exports[`options transform default mapping 1`] = `
<div>
  <div>
    <span>
      Default mapping
    </span>
  </div>
</div>
`;

exports[`whitespace allow newline only text node between elements 1`] = `
<pre>
  <span>
    Hello
  </span>
  

  <span>
    World
  </span>
</pre>
`;

exports[`whitespace allow whitespace only text nodes between elements 1`] = `
Array [
  <span>
    Hello
  </span>,
  " ",
  <span>
    World
  </span>,
]
`;

exports[`whitespace remove whitespace on table elements 1`] = `
<table>
  <tbody>
    <tr>
      <th>
        title
      </th>
    </tr>
    <tr>
      <td>
        entry
      </td>
    </tr>
  </tbody>
</table>
`;
